src/replay/consensus.zig

```diff
@@ -0,0 +1,1213 @@
+fn pushVote(
+    replay_tower: *ReplayTower,
+) !void {
+
+    // TODO Transaction generation to be implemented.
+    // Currently hardcoding to non voting transaction.
+    const vote_tx_result: GenerateVoteTxResult = .non_voting;
+
+    switch (vote_tx_result) {
+        .tx => |vote_tx| {
+            _ = &vote_tx;
+            // TODO to be implemented
+        },
+        .non_voting => {
+            replay_tower.markLastVoteTxBlockhashNonVoting();
+        },
+        else => {
+            // Do nothing
+        },
+    }
+}
+
+/// Processes a new root slot by updating various system components to reflect the new root.
+///
+/// - Validates the new root exists and has a hash.
+/// - Gets all slots rooted at the new root.
+/// - Updates ledger state with the new roots.
+/// - Cleans up progress map for non-existent slots.
+/// - Updates fork choice with the new root.
+///
+/// Analogous to [check_and_handle_new_root](https://github.com/anza-xyz/agave/blob/ccdcdbe9b6ff7dbd583d2101fe57b7cc41a6f863/core/src/replay_stage.rs#L4002)
+fn checkAndHandleNewRoot(
+    allocator: std.mem.Allocator,
+    ledger_result_writer: *LedgerResultWriter,
+    slot_tracker: *SlotTracker,
+    progress: *ProgressMap,
+    fork_choice: *ForkChoice,
+    new_root: Slot,
+) !void {
+    // get the root bank before squash.
+    if (slot_tracker.slots.count() == 0) return error.EmptySlotTracker;
+    const root_tracker = slot_tracker.get(new_root) orelse return error.MissingSlot;
+    const maybe_root_hash = root_tracker.state.hash.readCopy();
+    const root_hash = maybe_root_hash orelse return error.MissingHash;
+
+    const rooted_slots = try slot_tracker.parents(allocator, new_root);
+    defer allocator.free(rooted_slots);
+
+    try ledger_result_writer.setRoots(rooted_slots);
+
+    // Audit: The rest of the code maps to Self::handle_new_root in Agave.
+    slot_tracker.root = new_root;
+    // TODO
+    // - Prune program cache bank_forks.read().unwrap().prune_program_cache(new_root);
+    // - Extra operations as part of setting new root:
+    //   - cleare reward cache root_bank.clear_epoch_rewards_cache
+    //   - extend banks banks.extend(parents.iter());
+    //   - operations around snapshot_controller
+    //   - After setting a new root, prune the banks that are no longer on rooted paths self.prune_non_rooted(root, highest_super_majority_root);
+
+    // Update the progress map.
+    var to_remove = try std.ArrayListUnmanaged(Slot).initCapacity(
+        allocator,
+        progress.map.count(),
+    );
+    defer to_remove.deinit(allocator);
+
+    for (progress.map.keys()) |progress_slot| {
+        if (slot_tracker.get(progress_slot) == null) {
+            to_remove.appendAssumeCapacity(progress_slot);
+        }
+    }
+
+    for (to_remove.items) |key| {
+        _ = progress.map.swapRemove(key);
+    }
+
```

Comment:
Shouldn't this be updating the slot tracker here to remove these slots? In agave they call `BankForks::set_root` before updating the progress map. This removes all these slots from BankForks, which would be equivalent to us removing them from the slot tracker. Then you wouldn't need to iterate over to_remove for the progress map since it's already handled by the first loop. Also, when removing them from slot_tracker, that could produce the list of parents, so you don't need to do a separate call to `parents`

I wonder if we should just combine the progress map with the slot tracker.

Reply:
Part of the `BankForks::set_root` operation is noted as TODO here https://github.com/Syndica/sig/blob/38ed8d6da8c95ac9b81502c8fb463f70914e4f64/src/replay/consensus.zig#L385

> I wonder if we should just combine the progress map with the slot tracker.

Not the first time the idea would be broached...and maybe the combination can be renamed to `replay_state_tracker`? As I think there is still some uncertainty to the `slot tracker` name?

But maybe these are optimization that can be made post having it run on test net?

Reply:
> I can't remember... Probably got mistaken annotated as part of the things to be implemented later.

Looking at it again, it was mainly because the slot tracker currently does not have a way to list descendants and less so, because there is no implementation of `block_commitment_cache` from where `highest_super_majority_root` can be retrieved. See https://github.com/anza-xyz/agave/blob/1c776667cb7f9e4f20891c08c256e82136829465/runtime/src/bank_forks.rs#L591

> Also regarding the highlighted code for this thread. Why are you using an array list for this? Why not just remove the items directly?

[5de6505](https://github.com/Syndica/sig/pull/762/commits/5de6505489e03de7f081ebf82254f8b46d400032)

Reply:
[a3fa6a7](https://github.com/Syndica/sig/pull/762/commits/a3fa6a70a17c7921ee7cffd4aa37cdd2c7aaebae)

Reply:
> Also regarding the highlighted code for this thread. Why are you using an array list for this? Why not just remove the items directly?
> > [5de6505](https://github.com/Syndica/sig/pull/762/commits/5de6505489e03de7f081ebf82254f8b46d400032)

Revisiting [5de6505](https://github.com/Syndica/sig/pull/762/commits/5de6505489e03de7f081ebf82254f8b46d400032)

I am not 100% sure if it is safe to be removing from the same map being iterated.

Reply:
Yeah I knew how agave works, but my point was that I don't think we need it right now. If we do need it, then it can be done using ancestors, which provides the same information as descendants but it's easier to initialize.

> I am not 100% sure if it is safe to be removing from the same map being iterated.

It can be if you are careful. This should work:

```zig
var keys = map.keys();
var index: usize = 0;
while (index < keys.len) {
    if (keys[index] < root) {
        map.swapRemoveAt(index);
        keys = map.keys();
    } else {
        index += 1;
    }
}
```

Reply:
I went ahead and implemented it so we can get this merged asap. 031e0b8372db3c3aff24ddefcb6568b6dec19c78